--- /opt/freepbx/panel/index.php.orig	2008-07-23 20:33:40.000000000 +0200
+++ /opt/freepbx/panel/index.php	2008-07-23 23:12:37.000000000 +0200
@@ -15,11 +15,11 @@
 if ($amp_conf["AMPWEBADDRESS"] == "")
 	{$amp_conf["AMPWEBADDRESS"] = $_SERVER["HTTP_HOST"];}
 	
-if ($_SERVER["HTTP_HOST"] != $amp_conf["AMPWEBADDRESS"]) {
-	$proto = ((isset($_SERVER["HTTPS"]) && ($_SERVER["HTTPS"] == "on")) ? "https" : "http");
-	header("Location: ".$proto."://".$amp_conf["AMPWEBADDRESS"].$_SERVER["REQUEST_URI"]);
-	exit;
-}
+//if ($_SERVER["HTTP_HOST"] != $amp_conf["AMPWEBADDRESS"]) {
+//	$proto = ((isset($_SERVER["HTTPS"]) && ($_SERVER["HTTPS"] == "on")) ? "https" : "http");
+//	header("Location: ".$proto."://".$amp_conf["AMPWEBADDRESS"].$_SERVER["REQUEST_URI"]);
+//	exit;
+//}
 
 ?>
 <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
